### DISCUSSION TITLE
My buffer dont "work".

### DISCUSSION THREAD

**Discussion by ghost** [ocornut/imgui#3847] [Old/Archived discussions]
Hello. I'm trying to make a simple login system using ImGui. 

I've made 2 TextInputs, one aks for the usernam, the other one asks the password. But when i try to check their value, they don't seem to have any value stored inside. Look at the code:
(Also, i still learning C++. I'm sorry if this is a dumb question).

```
static char buf[128] = "";
static char buf2[128] = "";

 ImGui::Text(xorstr_("username:"));
        ImGui::InputText("##uname", buf, sizeof(buf), ImGuiInputTextFlags_EnterReturnsTrue);
        ImGui::Spacing();
        ImGui::Spacing();
        ImGui::Spacing();
        ImGui::Text(xorstr_("password:"));
        ImGui::InputText("##pwd", buf2, sizeof(buf), ImGuiInputTextFlags_EnterReturnsTrue);
        ImGui::Spacing();
        ImGui::Spacing();
        ImGui::Spacing();
        ImGui::SameLine(100.0f);

        if (ImGui::Button("Login", ImVec2(70, 25))) {
            if (!authed) {
                if (buf == "rando") {
                    authed = true;
                }
            }
        }

```

And when i type the username in the text field one (buf) and press the button, even if it has the value "rando" (the correct username), it don't login with "authed = true".

(authed is the variable that initialize the program menu once the login is successful.

Can someone help me?

**Comment by ghost** [comment]
I've also used IM_ARRAYSIZE(buf) instead of sizeof(buf), but still the same xD

**Comment by PathogenDavid** [comment]
> ```cpp
> if (buf == "rando") {
> ```

You can't compare `char` arrays like this in C/C++.

You either need to use [`strcmp`](https://www.cplusplus.com/reference/cstring/strcmp/):

```cpp
if (strcmp(buf, "rando") == 0) {
```

or concert one of the buffers to `std::string`, which does allow direct comparison:

```cpp
if (std::string(buf) == "rando") {
```

-----------------------

> ```cpp
> ImGui::InputText("##pwd", buf2, sizeof(buf), ImGuiInputTextFlags_EnterReturnsTrue);
> ```

You're using `sizeof(buf)` where you should've used `sizeof(buf2)`.It technically doesn't matter in this scenario since they're the same size, but this would turn into an annoying bug if they ever changed sizes.

-----------------------

> I've also used IM_ARRAYSIZE(buf) instead of sizeof(buf), but still the same xD

As you might've guessed by now, this wasn't the source of your problem so you could change it back. For `char`, `sizeof` and `IM_ARRAYSIZE` are the same, but it's a good habit since it clarifies the intent. (Number of bytes in the buffer vs number of characters in the buffer.)

-----------------------

As an aside, make sure you pay attention to your compiler warnings. Warnings almost always mean "What you're doing is technically valid C++ but it looks wrong." (At this point in your learning, you should generally assume the compiler is right and you're wrong.)

Both MSVC and Clang emit warnings for your code:

> warning C4130: '==': logical operation on address of string constant

> warning: array comparison always evaluates to false [-Wtautological-compare]

(GCC doesn't seem to warn for this unfortunately.)

-----------------------

> (Also, i still learning C++. I'm sorry if this is a dumb question).

No worries, we all gotta learn at some point! Strings in C/C++ have a bit of a learning curve compared to other languages.

Next time consider inspecting the contents of `buf` manually. In this scenario you should've seen that the buffer had the correct data, which would've given you a hint that your use of Dear ImGui was fine and that it was your comparison that was broken.

To do that, you could either print it to the console:

```cpp
// C-centric way that still works in C++:
printf("%s\n", buf); // Include <stdio.h>

// C++-centric way:
std::cout << buf << std::endl; // Include <iostream>

// Both techniques have the same result, it really boils down to which one you prefer.
// If you're working in a code base with other people, use whatever is being used elsewhere.
```

or you can also inspect the values in your debugger. How you do this depends on your development environment. With most IDEs, click the margin to the left of your code (and left of the line numbers) and a red circle will appear. This red circle represents a breakpoint. When you run your app, it will be paused when it his the breakpoint, allowing you to inspect the contents of variables and such.

If you place a breakpoint on your `if (!authed) {` line and run your app (by launching it from the IDE) and then click your "Login" button, the app will stop at that point. Switch back to your IDE and you'll be able to inspect program state and such. Usually you can hover over the variables in the editor or look for a pane labeled "Locals". Here's an example of what it looks like in Visual Studio:

![image](https://user-images.githubusercontent.com/278957/109376497-77766580-788a-11eb-9b8f-fa7929d7970f.png)

The box underlined with red is what appears when I hovered my mouse over `buf` and the red-boxed area is the locals window.

Some other useful tools visible here as well, such as the Call Stack view at the right-bottom and the Step Into, Step Over, Step Out buttons in the toolbar towards the right-top to the right of the Stop button.

For example, you could hit the Step Over button to see how your program flows through the code. (In this case, you'd see it'd always skip over the `if (buf == "rando") {` if statement since the comparison always fails.)

The Call Stack view basically shows you how your program got to this point and what function it will resume when the current one returns. In this screenshot you can see the current function is `GH3847` and it was called from `main`. (The other entries are weird internal things that run before `main`.) You can double-click on the entries to inspect variables in previous functions.

Debuggers are very powerful tools and are worth learning, but there's also no shame in `printf` debugging either. (For certain problems, `printf` debugging can be superior. As with all things, different tools are better than others at solving different problems.)

**Reply by ghost** [reply]
Wow. Thanks a lot! I did not knew the function `strcmp`, this saved my system. Also, i did knew that i had to debug something, i just don't knew what.
Again, thanks!


Sorry for taking too long for answering.

**Reply by PathogenDavid** [reply]
No problem! And no worries, it was a long response 😅
